### ISSUE TITLE
Question about data split

### ISSUE THREAD

**Issue by abstrcode** [RUCAIBox/RecBole#871]
It's not clear why data split needs to ensure that there are no intersection of items between train/valid/test set? Which piece of code reflect this point? Is this a benchmark setting? Besides, it seems no such restriction in the sequential recommendation methods as the history index is None.

**Comment by hyp1231** [comment]
Hi @abstrcode,
Sorry I didn't get what you mean.
Could you please give a more detailed explanation about the meaning of "intersection of items", or some examples?
Do you mean that "the same target item occurs in train/valid/test set in sequential recommendation"?

**Comment by abstrcode** [comment]
@hyp1231 "intersection" means there are common items between two sets (i.e., training set, validation set, test set)

**Comment by hyp1231** [comment]
It seems that RecBole doesn't make this restriction.

As for my side, I run `GRU4Rec` on `ml-1m`, and set breakpoint in `recbole.quick_start.quick_start.run_recbole` as following,
```python
...
# dataset splitting
train_data, valid_data, test_data = data_preparation(config, dataset)

import ipdb; ipdb.set_trace()

# model loading and initialization
model = get_model(config['model'])(config, train_data.dataset).to(config['device'])
logger.info(model)
...
```
Then I get the remapped ID of target items in train and valid set,
```python
ipdb> train_target_item = train_data.dataset.inter_feat['item_id'][train_data.dataset.target_index]
ipdb> train_target_item
tensor([  28,   32,   38,  ..., 1125,  255,  408])

ipdb> valid_target_item = valid_data.dataset.inter_feat['item_id'][valid_data.dataset.target_index]
ipdb> valid_target_item
tensor([  35,   83,  169,  ..., 1438,  656,  155])
```
Then I obtain the intersection of target items in train and valid set,
```python
ipdb> np.intersect1d(train_target_item, valid_target_item)
array([   1,    2,    3, ..., 3508, 3596, 3621])
ipdb> np.intersect1d(train_target_item, valid_target_item).shape
(1868,)
```

Thus it seems that there are intersection of items between train/valid/test set.
Could you provide an running example, where you found RecBole restricts the intersection between items in train/valid/test set?

**Comment by abstrcode** [comment]
@hyp1231 Thanks for clarification. My question is due to the evaluation setting in acceleration for top-k selection as claimed by "Our solution is to set the scores of training items to the negative infinity, and perform the full ranking over the entire item set without removing training items.". This setting indicates these is no intersection between test set and training/validation set.

**Comment by hyp1231** [comment]
@abstrcode Thanks, I think I finally understand. 😆 

> Our solution is to set the scores of training items to the negative infinity, and perform the full ranking over the entire item set without removing training items.

The above description is indeed a bit confusing. Actually, the "training items" here means those **items have interacted with a certain user historically**, and the training items are different while evaluating different users. In other words, evaluation of RecBole adopts a way of non-repeatable recommendation, currently.

We will support evaluation for both repeatable and non-repeatable recommendation in the future, which can be controlled by args.

**Comment by abstrcode** [comment]
@hyp1231 If the test item appears in the history behavior, then this type of evaluation does not work? Do you filter out the test item which appears in the training/validation set? Besides, it seems sequential recommendation does no utilize this setting as history_index = None.

**Comment by hyp1231** [comment]
Opps, I made a mistake! Yes, you are right, and thanks for the correction.

Following the common evaluation settings in the literatures:
* Only sequential recommendation in RecBole adopts **repeatable** recommendation, which means items exist in the historical behaviors will be evaluated in `full sort` predict.
* While others (general, context-aware, knowledge-based) adopts non-repeatable recommendation.
